### ISSUE TITLE
New component macro is dependent on a variable name 'root'

### ISSUE THREAD

**Issue by cofee-on-the-desk** [Relm4/Relm4#112]
The 'simple' example compiles, but when I change name of the variable 'root' to anything else, it doesn't.

```rust
  // Initialize the UI.
    fn init_parts(
        counter: Self::InitParams,
        window: &Self::Root, // changed 'root' to 'window'
        input: &mut Sender<Self::Input>,
        _output: &mut Sender<Self::Output>,
    ) -> ComponentParts<Self, Self::Widgets> {
```

Error:

```rust
error[E0425]: cannot find value `root` in this scope
  --> src/main.rs:25:9
   |
25 |         gtk::Window {
   |         ^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
```

**Comment by AaronErhardt** [comment]
That need to be addressed for sure. I've had the idea to add a syntax like this:

```rust
let widgets = view_output!(model, root);
```

Or we could use the function signature to figure out the names, but then still something like this would be necessary to be 100% correct:

```rust
let widgets = view_output!(model);
```

**Comment by cofee-on-the-desk** [comment]
I wonder if we could reduce the amound of boilerplate by stripping out the type info from the function parameters,

so this:

```rust
fn init_parts(
        counter: Self::InitParams,
        root: &Self::Root,
        input: &mut Sender<Self::Input>,
        _output: &mut Sender<Self::Output>,
    ) -> ComponentParts<Self, Self::Widgets> {
        let model = AppModel { counter };
        let widgets = view_output!();
        ComponentParts { model, widgets }
    }
```

turns into this:

```rust
init_parts! (counter, root, input, _output) {
        let model = AppModel { counter };
        let widgets = view_output!();
        ComponentParts { model, widgets }
}
```

This is not very rusty, but it looks nicer.

**Comment by cofee-on-the-desk** [comment]
Oh, I see, you've been suggesting the same idea in #99. So what do you think now?

**Comment by AaronErhardt** [comment]
I think, Rust will actually prevent you from using this syntax. It's weird but everything that isn't the direct input of a macro must be somewhat valid syntax. But in any case you're right and I will experiment with solutions to reduce the boilerplate.

I think the types could be reduces, too, if everything that isn't explicitly declared defaults to `()` unless it's something important like `Widgets`.

**Comment by mmstick** [comment]
I did get a syntax similar to this to work with macro_rules in my original ccs prototype.

**Comment by mmstick** [comment]
Rust permits it as long as the user provides `ident` arguments

```rs
$(#[$attr4:meta])*
fn init(
    $argsv:ident: $args:ty,
    $rootv:ident,
    $inputv:ident,
    $outputv:ident
) $init_view:block
```

**Comment by AaronErhardt** [comment]
I had a similar syntax planned for the [menu macro](https://aaronerhardt.github.io/blog/posts/announcing_relm4_v0.4/#menu-macro) but it didn't work. Rustc doesn't like something like `macro!(param) { other code }`, even if it's part of an attribute proc-macro.

The solution with `fn` could work, though. Yet I still have to test because, `macro_rules!` and proc macros are quite different.

**Comment by AaronErhardt** [comment]
We will have a problem with custom syntax and macros in Rust, it seems. Rust only likes custom syntax once you have at least an outer `macro! { input }` to put everything into.

```rust
/// Custom syntax NOT allowed
#[test]
struct Test(Custom) {
    test: u8,
}

/// Custom syntax allowed
test! {
    struct Test(Custom) {
        test: u8,
    }
}
```

This mean that attribute macros like `#[component]` don't allow custom syntax on their own. Yet once we use a normal proc-macro we might loose support for some tooling including cargo fmt if I'm not mistaken. Also, it's usually kind of nice that Rust catches invalid syntax in the trait impl before the macro has to do anything.

TL:DR we can't use custom syntax in attribute macros directly and it would be nice to keep the macro as attribute macro (to keep the code as natural as possible).

However, what we can do is to use valid Rust syntax for our advantage.

For example, we can do something like this:

```rust
    // Initialize the UI.
    #[args(params, root, input, output)]
    fn init_parts() {
        ...
    }
```

It'd definitely support cargo fmt and looks more Rust-ish, I think.

**Comment by mmstick** [comment]
`#[args(params, root, input, output)]` sounds like a good idea if it can work better with IDEs.

**Comment by AaronErhardt** [comment]
Should be fixed now. The macro looks the function parameters and the returned model to figure out the name of the model and to root.
